File: platform-services/search-service/app/kafka_consumer.py
import json
import logging
import os
import threading
import time

from kafka import KafkaConsumer
from kafka.errors import NoBrokersAvailable

logger = logging.getLogger(__name__)

KAFKA_BROKERS = os.environ.get("KAFKA_BROKERS", "localhost:9092")
TASKS_TOPIC = "task-service.tasks"
CONSUMER_GROUP = "search-service-tasks-indexer"


def task_event_to_doc(event: dict) -> dict:
    """Maps a task-service lifecycle event onto the tasks-<service> mapping (6.1).

    Deliberately whitelists fields rather than passing the event through
    as-is — task_id/service/event on the event are routing info, not doc
    content (task_id becomes the ES _id, service selects the index,
    event is unused after routing here).
    """
    return {
        "title": event.get("title"),
        "context": event.get("context"),
        "status": event.get("status"),
        "tags": event.get("tags") or [],
        "owner": event.get("owner"),
        "assignee_ids": event.get("assignee_ids") or [],
        "customer_id": event.get("customer_id"),
        "account_id": event.get("account_id"),
        "project_id": event.get("project_id"),
        "created_at": event.get("created_at"),
        "assigned_at": event.get("assigned_at"),
    }


def index_task_event(es, event: dict, service_index_name) -> None:
    """Upsert, not append.

    Every lifecycle transition (task.assigned, task.moved-to-review,
    task.reviewer-reassigned, task.approved, task.rejected) republishes
    the task's full current state, and _id = task_id (6.1.3) means this
    is always "replace with latest," never a partial patch — that's what
    makes "index on assign / update on reassign or reviewer-change" a
    single code path instead of one per event type.
    """
    task_id = event.get("task_id")
    service = event.get("service")
    if not task_id or not service:
        logger.warning("skipping task event missing task_id/service: %s", event)
        return
    es.index(index=service_index_name(service), id=task_id, document=task_event_to_doc(event))


def _build_consumer(brokers: str) -> KafkaConsumer:
    return KafkaConsumer(
        TASKS_TOPIC,
        bootstrap_servers=brokers.split(","),
        group_id=CONSUMER_GROUP,
        auto_offset_reset="earliest",
        value_deserializer=lambda raw: raw,
    )


def run_consumer_with_retry(es, service_index_name, brokers: str = KAFKA_BROKERS, stop_event: threading.Event = None) -> None:
    """Blocking loop — meant to run on its own thread (see start_kafka_consumer_thread),
    not the asyncio event loop, matching how the rest of this module already
    calls the sync Elasticsearch client directly. Kafka may not be reachable
    yet at search-service boot (container startup ordering) or the topic may
    not exist yet (task-service hasn't published anything) — both retry
    rather than crash the app, same posture as rustledger's connect-retry
    loop (domain-services/rustledger/src/kafka_consumer.rs).
    """
    stop_event = stop_event or threading.Event()
    while not stop_event.is_set():
        try:
            consumer = _build_consumer(brokers)
        except NoBrokersAvailable:
            logger.warning("kafka not reachable yet, retrying in 5s")
            time.sleep(5)
            continue

        try:
            for message in consumer:
                if stop_event.is_set():
                    break
                try:
                    event = json.loads(message.value)
                except (json.JSONDecodeError, TypeError):
                    logger.error("failed to parse task event, skipping: %r", message.value)
                    continue
                index_task_event(es, event, service_index_name)
        except Exception:
            logger.exception("kafka consumer error, retrying in 5s")
            time.sleep(5)
        finally:
            consumer.close()


def start_kafka_consumer_thread(es, service_index_name, brokers: str = KAFKA_BROKERS) -> tuple:
    """Starts run_consumer_with_retry on a daemon thread. Returns (thread, stop_event)
    so callers can signal a clean shutdown (see main.py's stop_kafka_consumer).
    """
    stop_event = threading.Event()
    thread = threading.Thread(
        target=run_consumer_with_retry,
        args=(es, service_index_name, brokers, stop_event),
        daemon=True,
        name="kafka-tasks-indexer",
    )
    thread.start()
    return thread, stop_event

File: platform-services/audit-service/events/kafka-consumer.js
// platform-services/audit-service/events/kafka-consumer.js
//
// One consumer group across two topics — task-service's existing
// task-service.tasks (a third independent consumer group there, alongside
// search-service's indexer and notification-service's notifier) and
// gofeeler's new gofeeler.sentiment (see domain-services/gofeeler/app/
// events/kafka.go for why that's a separate topic rather than riding on
// task-service.tasks: search-service's consumer does a blind full-state
// upsert on every message it sees there, and a sentiment.analyzed message
// shares none of those fields). Every row is written verbatim off
// whichever event arrives — no diffing against prior state, no lookup —
// the event name already carries the transition semantics.
const { Kafka } = require('kafkajs');
const { insertEvent } = require('../models/audit');

const TASK_TOPIC = 'task-service.tasks';
const SENTIMENT_TOPIC = 'gofeeler.sentiment';
const GROUP_ID = 'audit-service-events';

const kafka = new Kafka({
  clientId: 'audit-service',
  brokers: (process.env.KAFKA_BROKERS || 'microverse-kafka:9092').split(','),
});

function rowForTaskEvent(event, message) {
  return {
    taskId: event.task_id,
    service: event.service,
    event: event.event,
    status: event.status ?? null,
    owner: event.owner ?? null,
    assignee: (event.assignee_ids && event.assignee_ids[0]) || null,
    durationMs: null,
    occurredAt: new Date(Number(message.timestamp)),
  };
}

function rowForSentimentEvent(event) {
  return {
    taskId: event.task_id,
    service: event.service,
    event: event.event,
    status: null,
    owner: null,
    assignee: null,
    durationMs: event.duration_ms ?? null,
    occurredAt: event.analyzed_at ? new Date(event.analyzed_at) : new Date(),
  };
}

async function handleMessage(topic, event, message) {
  const row = topic === SENTIMENT_TOPIC ? rowForSentimentEvent(event) : rowForTaskEvent(event, message);
  await insertEvent(row);
}

// Same posture as notification-service's kafka-consumer.js: Kafka may not
// be reachable yet at boot, or the consumer can crash mid-stream — both
// cases retry after 5s rather than bringing the service down.
function startConsumer() {
  const consumer = kafka.consumer({ groupId: GROUP_ID });

  const retry = (reason, err) => {
    console.error(`${reason}, retrying in 5s:`, err.message);
    consumer
      .disconnect()
      .catch(() => {})
      .finally(() => setTimeout(() => startConsumer(), 5000));
  };

  consumer.on(consumer.events.CRASH, ({ payload }) => retry('kafka consumer crashed', payload.error));

  (async () => {
    try {
      await consumer.connect();
      await consumer.subscribe({ topics: [TASK_TOPIC, SENTIMENT_TOPIC], fromBeginning: true });
      await consumer.run({
        eachMessage: async ({ topic, message }) => {
          let event;
          try {
            event = JSON.parse(message.value.toString());
          } catch (err) {
            console.error('failed to parse audit event, skipping:', err.message);
            return;
          }
          try {
            await handleMessage(topic, event, message);
          } catch (err) {
            console.error(`error handling ${event.event} for task ${event.task_id}:`, err.message);
          }
        },
      });
    } catch (err) {
      retry('kafka consumer failed to start', err);
    }
  })();
}

module.exports = { startConsumer };
